kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/partition/DynamicPartitioner.java

```diff
@@ -0,0 +1,256 @@
+      // decrease batch size back to the size that we actually checked.
+      decreaseBatchSize();
+      logger.debug("Best batch size found (all elements): {}", currentBatchSize);
+    }
+  }
+
+  /**
+   * writeAll intended for a generally stable currentBatchSize.
+   *
+   * <p>Write all the given elements to BigQuery in the smallest possible number of batches, as
+   * evenly as possible.
+   *
+   * <p>If we encounter a partitioning-related error, we will immediately lower the batch size and
+   * try again.
+   *
+   * <p>Every {@link #CONT_SUCCESS_COUNT_BUMP} calls, if there have been no errors, we will bump up
+   * the batch size.
+   */
+  private void establishedWriteAll(TableId table,
+                                   List<InsertAllRequest.RowToInsert> elements,
+                                   String topic,
+                                   Set<Schema> schemas) throws BigQueryConnectException,
+                                                               InterruptedException {
+    int currentIndex = 0;
+    while (currentIndex < elements.size()) {
+      try {
+        // very similar to equalPartitioner
+        // handle case where no partitioning is necessary:
+        if (elements.size() <= currentBatchSize) {
+          writer.writeRows(table, elements, topic, schemas);
+          currentIndex = elements.size();
+          return;
+        }
+
+        int numPartitions = (int)Math.ceil(elements.size() / (currentBatchSize * 1.0));
+        int minBatchSize = elements.size() / numPartitions;
+        int spareRows = elements.size() % numPartitions;
+
+        for (int partition = 0; partition < numPartitions; partition++) {
+          // the first spareRows partitions have an extra row in them.
+          int batchSize = partition < spareRows ? minBatchSize + 1 : minBatchSize;
+          int endIndex = currentIndex + batchSize;
+          writer.writeRows(table, elements.subList(currentIndex, endIndex), topic, schemas);
+          currentIndex = endIndex;
+        }
+      } catch (BigQueryException exception) {
+        if (isPartitioningError(exception)) {
+          // immediately decrease batch size and try again with remaining elements.
+          decreaseBatchSize();
+          contSuccessCount = 0;
+          logger.debug("Partition error during establishedWriteAll, reducing batch size to {}",
+                       currentBatchSize);
+          continue;
+        } else {
+          throw new BigQueryConnectException(
+              String.format("Failed to write to BigQuery table %s", table),
+              exception);
+        }
+      }
+    }
+
+    contSuccessCount++;
+    if (contSuccessCount >= CONT_SUCCESS_COUNT_BUMP) {
+      logger.debug("{} successful establishedWriteAlls in a row, increasing batchSize to {}",
+                   contSuccessCount, currentBatchSize);
+      increaseBatchSize();
+    }
+  }
+
+  private boolean isPartitioningError(BigQueryException exception) {
```

Comment:
We aren't handling rateLimitExceeded anywhere atm.  I think the correct place to handle that would be the BQWriter, but I don't think that change should be part of this RB.

If I'm interpreting correctly It sounds like you want to also catch 400s that are "invalid" (because that's what the spec says it will/should throw)?  I can do that.